File: CodeGame/java/Actor.java
package game;


public abstract class Actor {
    /**
     * The ActorType enum represents the different types of actors that can exist in the game.
     * Each actor type has a specific role or function within the game.
     * 
     * The ActorType enum specifies whether the actor is a player, obstacle, destination, or item.
     * - PLAYER: Represents a player actor.
     * - OBSTACLE: Represents an obstacle actor.
     * - DESTINATION: Represents a destination actor.
     * - ITEM: Represents a item actor.
     */
    public enum ActorType {
        PLAYER,
        OBSTACLE,
        DESTINATION,
        ITEM
    }

    /**
     * The ActorDirectionInWorld enum represents the four cardinal directions.
     * It is used to specify the direction in which an actor is facing or moving.
     * 
     * The ActorDirectionInWorld enum specifies the direction the actor is facing.
     * - NORTH: Facing north.
     * - EAST: Facing east.
     * - SOUTH: Facing south.
     * - WEST: Facing west.
     */
    public enum ActorDirectionInWorld {
        NORTH,
        EAST,
        SOUTH,
        WEST
    }

    /**
     * The ActorDirection enum represents the two possible directions an actor can turn.
     * It is used to specify the direction in which an actor can turn.
     * 
     * The ActorDirection enum specifies the direction the actor can turn.
     * - LEFT: Turning left.
     * - RIGHT: Turning right.
     */
    public enum ActorDirection {
        LEFT,
        RIGHT
    }

    protected int x;
    protected int y;
    protected ActorType actorType;
    protected ActorDirectionInWorld actorDirectionInWorld;
    protected World world;
    protected boolean gameError = false;

    /**
     * Constructs an Actor with the specified position, direction, type, and world.
     *
     * @param x the x-coordinate of the actor
     * @param y the y-coordinate of the actor
     * @param actorDirectionInWorld the direction the actor is facing
     * @param actorType the type of the actor
     * @param world the world in which the actor exists
     */
    public Actor(int x, int y, ActorDirectionInWorld actorDirectionInWorld, ActorType actorType, World world) {
        this.x = x;
        this.y = y;
        this.actorDirectionInWorld = actorDirectionInWorld;
        this.actorType = actorType;
        this.world = world;
    }

    /**
     * Defines the action that this actor will perform.
     * This method should be implemented by subclasses to specify the behavior of the actor.
     */
    public abstract void act();

    /**
     * Moves the actor a specified distance in the current direction.
     * 
     * @param distance the distance to move the actor
     */
    public void move(int distance) {
        if (gameError) {
            return;
        }

        if (checkWorldBounds()) {
            SystemOutput.getInstance().outputInformation("World boundary detected.");
            SystemOutput.getInstance().outputWarning();
            gameError = true;
            return;
        }

        if (checkObstacle()) {
            SystemOutput.getInstance().outputInformation("Obstacle detected in front.");
            return;
        }

        int newX = x;
        int newY = y;

        switch (actorDirectionInWorld) {
            case NORTH:
                newY -= distance;
                break;
            case EAST:
                newX += distance;
                break;
            case SOUTH:
                newY += distance;
                break;
            case WEST:
                newX -= distance;
                break;
        }

        if (world != null) {
            world.moveObject((Player) this, newX, newY);
            x = newX;
            y = newY;
        } else {
            System.err.println("Actor.java - move: World instance is not reachable.");
        }
    }

    /**
     * Turns the actor to the specified direction.
     *
     * @param direction The direction to turn the actor to.
     */
    public void turn(ActorDirection direction) {
        if (gameError) {
            return;
        }

        if (direction == ActorDirection.LEFT) {
            if (actorDirectionInWorld == ActorDirectionInWorld.NORTH) {
                actorDirectionInWorld = ActorDirectionInWorld.WEST;
            } else if (actorDirectionInWorld == ActorDirectionInWorld.WEST) {
                actorDirectionInWorld = ActorDirectionInWorld.SOUTH;
            } else if (actorDirectionInWorld == ActorDirectionInWorld.SOUTH) {
                actorDirectionInWorld = ActorDirectionInWorld.EAST;
            } else if (actorDirectionInWorld == ActorDirectionInWorld.EAST) {
                actorDirectionInWorld = ActorDirectionInWorld.NORTH;
            }
        } else if (direction == ActorDirection.RIGHT) {
            if (actorDirectionInWorld == ActorDirectionInWorld.NORTH) {
                actorDirectionInWorld = ActorDirectionInWorld.EAST;
            } else if (actorDirectionInWorld == ActorDirectionInWorld.EAST) {
                actorDirectionInWorld = ActorDirectionInWorld.SOUTH;
            } else if (actorDirectionInWorld == ActorDirectionInWorld.SOUTH) {
                actorDirectionInWorld = ActorDirectionInWorld.WEST;
            } else if (actorDirectionInWorld == ActorDirectionInWorld.WEST) {
                actorDirectionInWorld = ActorDirectionInWorld.NORTH;
            }
        }

        SystemOutput.getInstance().outputTurn(getDirectionString());
    }

    /**
     * Checks if there is an obstacle in the direction the actor is facing.
     *
     * @return true if there is an obstacle in the direction the actor is facing, false otherwise.
     */
    public boolean checkObstacle() {
        switch (actorDirectionInWorld) {
            case NORTH:
                return world.getObstacles(x, y - 1).size() > 0;
            case EAST:
                return world.getObstacles(x + 1, y).size() > 0;
            case SOUTH:
                return world.getObstacles(x, y + 1).size() > 0;
            case WEST:
                return world.getObstacles(x - 1, y).size() > 0;
            default:
                return false;
        }
    }

    /**
     * Checks if the actor is within the bounds of the world.
     * 
     * @return true if the actor is within the bounds of the world, false otherwise.
     */
    public boolean checkWorldBounds() {
        if (world == null) {
            System.err.println("Actor.java - checkWorldBounds: World instance is not reachable.");
            return false;
        }

        switch (actorDirectionInWorld) {
            case NORTH:
                return y - 1 < 0;
            case EAST:
                return x + 1 >= world.getWidth();
            case SOUTH:
                return y + 1 >= world.getHeight();
            case WEST:
                return x - 1 < 0;
            default:
                return false;
        }
    }

    /**
     * Returns the x-coordinate of the actor.
     *
     * @return the x-coordinate
     */
    public int getX() {
        return x;
    }

    /**
     * Returns the y-coordinate of the actor.
     *
     * @return the y-coordinate
     */
    public int getY() {
        return y;
    }

    /**
     * Retrieves the current direction of the actor.
     *
     * @return the direction of the actor as an {@link ActorDirectionInWorld} enum.
     */
    public ActorDirectionInWorld getDirection() {
        return actorDirectionInWorld;
    }

    /**
     * Returns the string representation of the actor's direction.
     *
     * @return A string representing the direction of the actor. Possible values are:
     *         "NORTH", "EAST", "SOUTH", "WEST". If the direction is not recognized,
     *         an empty string is returned.
     */
    public String getDirectionString() {
        switch (actorDirectionInWorld) {
            case NORTH:
                return "NORTH";
            case EAST:
                return "EAST";
            case SOUTH:
                return "SOUTH";
            case WEST:
                return "WEST";
            default:
                return "";
        }
    }

    /**
     * Retrieves the type of the actor.
     *
     * @return the type of the actor as an {@link ActorType}.
     */
    public ActorType getType() {
        return actorType;
    }

    /**
     * Returns the string representation of the actor type.
     *
     * @return a string representing the actor type, which can be "PLAYER", "OBSTACLE",
     *         "DESTINATION", "ITEM", or an empty string if the actor type is not recognized.
     */
    public String getActorTypeString() {
        switch (actorType) {
            case PLAYER:
                return "PLAYER";
            case OBSTACLE:
                return "OBSTACLE";
            case DESTINATION:
                return "DESTINATION";
            case ITEM:
                return "ITEM";
            default:
                return "";
        }
    }

    /**
     * Returns the current game error status.
     *
     * @return true if there is a game error, false otherwise.
     */
    public boolean getGameError() {
        return gameError;
    }
}

File: CodeGame/py/Actor.py
from enum import Enum
from SystemOutput import SystemOutput

class ActorType(Enum):
    """
    Enum class representing different types of actors in the game.

    Attributes:
        PLAYER (int): Represents the player character.
        OBSTACLE (int): Represents an obstacle in the game.
        DESTINATION (int): Represents the destination or goal in the game.
        ITEM (int): Represents an item that can be collected or used in the game.
    """
    PLAYER = 1
    OBSTACLE = 2
    DESTINATION = 3
    ITEM = 4

class ActorDirectionInWorld(Enum):
    """
    Enum class representing the four cardinal directions an actor can face.

    Attributes:
        NORTH (int): Represents the north direction with a value of 1.
        EAST (int): Represents the east direction with a value of 2.
        SOUTH (int): Represents the south direction with a value of 3.
        WEST (int): Represents the west direction with a value of 4.
    """
    NORTH = 1
    EAST = 2
    SOUTH = 3
    WEST = 4

class ActorDirection(Enum):
    """
    Enum class representing the two possible directions an actor can turn.

    Attributes:
        LEFT (int): Represents the left direction with a value of 1.
        RIGHT (int): Represents the right direction with a value of 2.
    """
    LEFT = 1
    RIGHT = 2

class Actor:
    def __init__(self, x, y, direction, actor_type, world):
        """
        Initialize an Actor object.

        Args:
            x (int): The x-coordinate of the actor.
            y (int): The y-coordinate of the actor.
            direction (str): The direction the actor is facing.
            actor_type (str): The type of the actor.
            world (World): The world in which the actor exists.
        """
        self.x = x
        self.y = y
        self.actor_direction = direction
        self.actor_type = actor_type
        self.world = world
        self.game_error = False

    def move(self, distance):
        """
        Moves the actor a specified distance in the current direction.

        Parameters:
        distance (int): The distance to move the actor.

        Returns:
        None

        Behavior:
        - If a game error has occurred, the method returns immediately.
        - If the actor is at the world boundary, outputs a warning and sets the game error flag.
        - If there is an obstacle in front of the actor, outputs an information message and does not move the actor.
        - Updates the actor's position based on the current direction (NORTH, EAST, SOUTH, WEST).
        - If the world instance is available, moves the actor within the world and updates the actor's coordinates.
        - If the world instance is not reachable, prints an error message.
        """
        if self.game_error:
            return

        if self.check_world_bounds():
            SystemOutput.get_instance().output_information("World boundary detected.")
            SystemOutput.get_instance().output_warning()
            self.game_error = True
            return

        if self.check_obstacle():
            SystemOutput.get_instance().output_information("Obstacle detected in front.")
            return

        new_x = self.x
        new_y = self.y

        if self.actor_direction == ActorDirectionInWorld.NORTH:
            new_y -= distance
        elif self.actor_direction == ActorDirectionInWorld.EAST:
            new_x += distance
        elif self.actor_direction == ActorDirectionInWorld.SOUTH:
            new_y += distance
        elif self.actor_direction == ActorDirectionInWorld.WEST:
            new_x -= distance

        if self.world:
            self.world.move_object(self, new_x, new_y)
            self.x = new_x
            self.y = new_y
        else:
            print("Actor.py - move: World instance is not reachable.")

    def turn(self, direction):
        """
        Changes the direction of the actor.

        Args:
            direction (ActorDirection): The new direction for the actor.
        """
        if self.game_error:
            return

        if (direction == ActorDirection.LEFT):
            if (self.actor_direction == ActorDirectionInWorld.NORTH):
                self.actor_direction = ActorDirectionInWorld.WEST
            elif (self.actor_direction == ActorDirectionInWorld.WEST):
                self.actor_direction = ActorDirectionInWorld.SOUTH
            elif (self.actor_direction == ActorDirectionInWorld.SOUTH):
                self.actor_direction = ActorDirectionInWorld.EAST
        elif (direction == ActorDirection.RIGHT):
            if (self.actor_direction == ActorDirectionInWorld.NORTH):
                self.actor_direction = ActorDirectionInWorld.EAST
            elif (self.actor_direction == ActorDirectionInWorld.EAST):
                self.actor_direction = ActorDirectionInWorld.SOUTH
            elif (self.actor_direction == ActorDirectionInWorld.SOUTH):
                self.actor_direction = ActorDirectionInWorld.WEST

        SystemOutput.get_instance().output_turn(self.get_direction_string())

    def check_obstacle(self):
        """
        Checks if there is an obstacle in the direction the actor is facing.

        Returns:
            bool: True if there is an obstacle in the direction the actor is facing, False otherwise.
        """
        if self.actor_direction == ActorDirectionInWorld.NORTH:
            return len(self.world.get_obstacles(self.x, self.y - 1)) > 0
        elif self.actor_direction == ActorDirectionInWorld.EAST:
            return len(self.world.get_obstacles(self.x + 1, self.y)) > 0
        elif self.actor_direction == ActorDirectionInWorld.SOUTH:
            return len(self.world.get_obstacles(self.x, self.y + 1)) > 0
        elif self.actor_direction == ActorDirectionInWorld.WEST:
            return len(self.world.get_obstacles(self.x - 1, self.y)) > 0
        return False

    def check_world_bounds(self):
        """
        Checks if the actor is within the bounds of the world.

        This method verifies if the actor's current position, based on its direction,
        is within the valid boundaries of the world. It returns False if the world
        instance is not reachable.

        Returns:
            bool: True if the actor is within the world bounds, False otherwise.
        """
        if not self.world:
            print("Actor.py - check_world_bounds: World instance is not reachable.")
            return False

        if self.actor_direction == ActorDirectionInWorld.NORTH:
            return self.y - 1 < 0
        elif self.actor_direction == ActorDirectionInWorld.EAST:
            return self.x + 1 >= self.world.get_width()
        elif self.actor_direction == ActorDirectionInWorld.SOUTH:
            return self.y + 1 >= self.world.get_height()
        elif self.actor_direction == ActorDirectionInWorld.WEST:
            return self.x - 1 < 0
        return False

    def get_x(self):
        """
        Returns the x-coordinate of the actor.

        Returns:
            int: The x-coordinate of the actor.
        """
        return self.x

    def get_y(self):
        """
        Returns the y-coordinate of the actor.

        Returns:
            int: The y-coordinate of the actor.
        """
        return self.y

    def get_direction(self):
        """
        Returns the current direction of the actor.

        Returns:
            str: The direction in which the actor is currently facing.
        """
        return self.actor_direction

    def get_direction_string(self):
        """
        Returns the string representation of the actor's current direction.

        The direction is determined by the `actor_direction` attribute, which is
        an instance of the `ActorDirection` enum. The possible return values are:
        - "NORTH" if the direction is `ActorDirectionInWorld.NORTH`
        - "EAST" if the direction is `ActorDirectionInWorld.EAST`
        - "SOUTH" if the direction is `ActorDirectionInWorld.SOUTH`
        - "WEST" if the direction is `ActorDirectionInWorld.WEST`

        Returns:
            str: The string representation of the actor's current direction, or
            an empty string if the direction is not recognized.
        """
        if self.actor_direction == ActorDirectionInWorld.NORTH:
            return "NORTH"
        elif self.actor_direction == ActorDirectionInWorld.EAST:
            return "EAST"
        elif self.actor_direction == ActorDirectionInWorld.SOUTH:
            return "SOUTH"
        elif self.actor_direction == ActorDirectionInWorld.WEST:
            return "WEST"
        return ""

    def get_type(self):
        """
        Returns the type of the actor.

        Returns:
            str: The type of the actor.
        """
        return self.actor_type

    def get_actor_type_string(self):
        """
        Returns the string representation of the actor type.

        Returns:
            str: The string representation of the actor type. Possible values are:
                 "PLAYER", "OBSTACLE", "DESTINATION", "ITEM", or an empty string if
                 the actor type does not match any known types.
        """
        if self.actor_type == ActorType.PLAYER:
            return "PLAYER"
        elif self.actor_type == ActorType.OBSTACLE:
            return "OBSTACLE"
        elif self.actor_type == ActorType.DESTINATION:
            return "DESTINATION"
        elif self.actor_type == ActorType.ITEM:
            return "ITEM"
        return ""

    def get_game_error(self):
        """
        Retrieve the current game error.

        Returns:
            str: The current game error message.
        """
        return self.game_error
